tesseract_motion_planners/src/core/utils.cpp

```diff
@@ -40,6 +40,150 @@ TESSERACT_COMMON_IGNORE_WARNINGS_POP
+  tesseract_common::VectorIsometry3d poses;
+
+  tesseract_environment::StateSolver::Ptr state_solver = env->getStateSolver();
+  tesseract_environment::EnvState::ConstPtr state = env->getCurrentState();
+  if (isCompositeInstruction(instruction))
+  {
+    const auto* ci = instruction.cast_const<CompositeInstruction>();
+
+    // Assume all the plan instructions have the same manipulator as the composite
+    assert(!ci->getManipulatorInfo().empty());
+    const ManipulatorInfo& composite_mi = ci->getManipulatorInfo();
+
+    auto composite_mi_fwd_kin = env->getManipulatorManager()->getFwdKinematicSolver(composite_mi.manipulator);
+    if (composite_mi_fwd_kin == nullptr)
+      throw std::runtime_error("toToolpath: Manipulator: " + composite_mi.manipulator + " does not exist!");
+
+    const std::string& tip_link = composite_mi_fwd_kin->getTipLinkName();
+
+    std::vector<std::reference_wrapper<const Instruction>> fi = tesseract_planning::flatten(*ci, planFilter);
+    if (fi.empty())
+      fi = tesseract_planning::flatten(*ci, moveFilter);
+
+    for (const auto& i : fi)
+    {
+      ManipulatorInfo manip_info;
+
+      // Check for updated manipulator information and get waypoint
+      Waypoint wp = NullWaypoint();
+      if (isPlanInstruction(i.get()))
+      {
+        const auto* pi = i.get().cast_const<PlanInstruction>();
+        manip_info = composite_mi.getCombined(pi->getManipulatorInfo());
+        wp = pi->getWaypoint();
+      }
+      else if (isMoveInstruction(i.get()))
+      {
+        const auto* mi = i.get().cast_const<MoveInstruction>();
+        manip_info = composite_mi.getCombined(mi->getManipulatorInfo());
+        wp = mi->getWaypoint();
+      }
+
+      // Extract TCP
+      Eigen::Isometry3d tcp = env->findTCP(manip_info);
+
+      if (isStateWaypoint(wp))
+      {
+        const auto* swp = wp.cast_const<StateWaypoint>();
+        assert(static_cast<long>(swp->joint_names.size()) == swp->position.size());
+        tesseract_environment::EnvState::Ptr state = state_solver->getState(swp->joint_names, swp->position);
+        poses.push_back(state->link_transforms[tip_link] * tcp);
+      }
+      else if (isJointWaypoint(wp))
+      {
+        const auto* jwp = wp.cast_const<JointWaypoint>();
+        assert(static_cast<long>(jwp->joint_names.size()) == jwp->size());
+        tesseract_environment::EnvState::Ptr state = state_solver->getState(jwp->joint_names, *jwp);
+        poses.push_back(state->link_transforms[tip_link] * tcp);
+      }
+      else if (isCartesianWaypoint(wp))
+      {
+        const auto* cwp = wp.cast_const<CartesianWaypoint>();
+        if (manip_info.working_frame.empty())
+        {
+          poses.push_back(*cwp);
+        }
+        else
+        {
+          poses.push_back(state->link_transforms.at(manip_info.working_frame) * (*cwp));
+        }
+      }
+      else
+      {
+        CONSOLE_BRIDGE_logError("toToolpath: Unsupported Waypoint Type!");
```

Comment:
Would we want a hard failure (exception) here instead? Or is this function primarily intended for visualization purposes?

Reply:
I think an exception is good.